File: backend/spring-boot/src/main/java/org/bugzkit/api/auth/jwt/util/JwtUtil.java
package org.bugzkit.api.auth.jwt.util;

import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import java.time.Instant;
import java.util.Set;
import java.util.stream.Collectors;
import org.bugzkit.api.user.payload.dto.RoleDTO;

public class JwtUtil {
  private static final String BEARER = "Bearer ";

  private JwtUtil() {}

  public static void verify(String token, String secret, JwtPurpose purpose) {
    JWT.require(JwtUtil.getAlgorithm(secret))
        .withClaim("purpose", purpose.name())
        .build()
        .verify(token);
  }

  public static Algorithm getAlgorithm(String secret) {
    return Algorithm.HMAC512(secret.getBytes());
  }

  public static String removeBearer(String token) {
    return token.replace(BEARER, "");
  }

  public static boolean isBearer(String token) {
    return token.startsWith(BEARER);
  }

  public static Long getUserId(String token) {
    return Long.parseLong(JWT.decode(token).getIssuer());
  }

  public static Set<RoleDTO> getRoleDTOs(String token) {
    final var roles = JWT.decode(token).getClaim("roles").asList(String.class);
    return roles.stream().map(RoleDTO::new).collect(Collectors.toSet());
  }

  public static Instant getIssuedAt(String token) {
    return JWT.decode(token).getIssuedAtAsInstant();
  }

  public enum JwtPurpose {
    ACCESS_TOKEN,
    REFRESH_TOKEN,
    VERIFY_EMAIL_TOKEN,
    RESET_PASSWORD_TOKEN
  }
}

File: backend/spring-boot-kotlin/src/main/kotlin/org/bootstrapbugz/api/auth/jwt/util/JwtUtil.kt
package org.bootstrapbugz.api.auth.jwt.util

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import java.time.Instant
import org.bootstrapbugz.api.user.payload.dto.RoleDTO

object JwtUtil {
  private const val BEARER = "Bearer "

  fun verify(token: String, secret: String, purpose: JwtPurpose) {
    JWT.require(getAlgorithm(secret)).withClaim("purpose", purpose.name).build().verify(token)
  }

  fun getAlgorithm(secret: String): Algorithm {
    return Algorithm.HMAC512(secret.toByteArray())
  }

  fun removeBearer(token: String): String {
    return token.replace(BEARER, "")
  }

  fun isBearer(token: String): Boolean {
    return token.startsWith(BEARER)
  }

  fun getUserId(token: String?): Long {
    return JWT.decode(token).issuer.toLong()
  }

  fun getRoleDTOs(token: String): Set<RoleDTO> {
    val roles = JWT.decode(token).getClaim("roles").asList(String::class.java)
    return roles.map { RoleDTO(it) }.toSet()
  }

  fun getIssuedAt(token: String): Instant {
    return JWT.decode(token).issuedAtAsInstant
  }

  enum class JwtPurpose {
    ACCESS_TOKEN,
    REFRESH_TOKEN,
    VERIFY_EMAIL_TOKEN,
    RESET_PASSWORD_TOKEN
  }
}
